source/kaleidic/lubeck2.d

```diff
@@ -2044,6 +2046,55 @@ unittest
     assert(equal!approxEqual(res.eigenvalues, eigenvalues));
 }
 
+/++
+Computes Moore-Penrose pseudoinverse of matrix.
+
+Params:
+    matrix = Input `M x N` matrix.
+    tolerance = The computation is based on SVD and any singular values less than tolerance are treated as zero.
+Returns: Moore-Penrose pseudoinverse matrix
++/
+@safe pure @nogc nothrow
+Slice!(RCI!T, 2)
+    pinv(T, SliceKind kind)(Slice!(const(T)*, 2, kind) matrix, double tolerance = double.nan)
+{
+    import mir.algorithm.iteration: find, each;
+    import std.math: nextUp;
+
+    auto svd = svd(matrix);
+    if (tolerance != tolerance)
+    {
+        auto n = svd.sigma.front;
+        auto eps = n.nextUp - n;
+        tolerance = max(matrix.length!0, matrix.length!1) * eps;
+    }
+    auto st = svd.sigma.find!(a => !(a >= tolerance));
+    static if (is(typeof(st) : sizediff_t))
+        alias si = st;
+    else
+        auto si = st[0];
+    auto s = svd.sigma[0 .. $ - si];
+    s.each!"a = 1 / a";
```

Comment:
I haven't tried this code or anything, but I would definitely want to make sure this line shouldn't be something like `s.each!"a = 1.0 / a"`

Reply:
`1 / ` is fine. It can be `1f / ` but `1.0 / `. The last one will trigger `float` to `double` conversion.

Reply:
I agree `1f` is better. I was a little concerned because the function signature doesn't constrain `T` to be floating point (though I imagine some of the other functions called make that restriction, it isn't immediately obvious from looking at this one if that is the case).